colossus/src/main/scala/colossus/protocols/http/HttpRequestParser.scala

```diff
@@ -25,13 +25,152 @@ object HttpRequestParser {
+ * read
+ */
+class HttpHeadParser extends Parser[HeadResult]{
+
+  class RBuilder {
+    var method: String = ""
+    var path: String = ""
+    var version: String = ""
+    var headers: List[(String, String)] = Nil
+    var contentLength: Option[Int] = None
+    var transferEncoding: Option[String] = None
+    var body: Option[ByteString] = None
+
+    def addHeader(rawName: String, rawValue: String) {
+      val name = rawName.toLowerCase
+      val value = rawValue.trim
+      headers = (name, value) :: headers
+      if (name == HttpHeaders.ContentLength) {
+        contentLength = Some(value.toInt)
+      } else if (name == HttpHeaders.TransferEncoding) {
+        transferEncoding = Some(value)
+      }
+    }
+
+    def build: HeadResult = {
+      val r = HeadResult(HttpHead(HttpMethod(method), path, HttpVersion(version), headers), contentLength, transferEncoding)
+      reset()
+      r
+    }
+
+    def reset() {
+      headers = Nil
+      contentLength = None
+      transferEncoding = None
+      body = None
+    }
+  }
+  var requestBuilder = new RBuilder
+  var headerState = 0 //incremented when parsing \r\n\r\n
+
+
+  trait MiniParser {
+    def parse(c: Char)
+    def end()
+  }
+
+  class FirstLineParser extends MiniParser {
+    val STATE_METHOD  = 0
+    val STATE_PATH    = 1
+    val STATE_VERSION = 2
+    var state = STATE_METHOD
+    val builder = new StringBuilder
+    def parse(c: Char) {
+      if (c == ' ') {
+        val res = builder.toString
+        builder.setLength(0)
+        state match {
+          case STATE_METHOD => {
+            requestBuilder.method = res
+            state = STATE_PATH
+          }
+          case STATE_PATH   => {
+            requestBuilder.path = res
+            state = STATE_VERSION
+          }
+          case _ => {
+            throw new ParseException("invalid content in header first line")
+          }
+        }
+      } else {
+        builder.append(c)
+      }
+    }
+    def end() {
+      val res = builder.toString
+      builder.setLength(0)
+      requestBuilder.version = res
+      state = STATE_METHOD
+    }
+  }
+
+  class HeaderParser extends MiniParser {
+    val STATE_KEY   = 0
+    val STATE_VALUE = 1
+    var state = STATE_KEY
+    val builder = new StringBuilder
+    var builtKey = ""
+    def parse(c: Char) {
+      if (c == ':' && state == STATE_KEY) {
+        builtKey = builder.toString
+        builder.setLength(0)
+        state = STATE_VALUE
+      } else {
+        builder.append(c)
+      }
+    }
+    def end() {
+      requestBuilder.addHeader(builtKey, builder.toString)
+      builder.setLength(0)
+      state = STATE_KEY
+    }
+  }
+
+  val fparser = new FirstLineParser
+  val hparser = new HeaderParser
+        
+  var currentParser: MiniParser = fparser
+
+
+  def parse(d: DataBuffer): Option[HeadResult] = {
+    while (d.hasUnreadData) {
+      val b = d.next.toChar
+      if (b == '\r') {
+        headerState += 1
```

Comment:
This makes it a bit hard to follow, ie: using magic numbers vs something else a bit more descriptive.